jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt

```diff
@@ -56,3 +59,74 @@ class SamExecutable : ExecutableType<SemVer>, AutoResolvable, Validatable {
         return Paths.get(path)
     }
 }
+
+fun GeneralCommandLine.samBuildCommand(
+    environmentVariables: Map<String, String>,
+    templatePath: Path,
+    buildDir: Path,
+    useContainer: Boolean
+) = this.apply {
+    withEnvironment(environmentVariables)
+    withWorkDirectory(templatePath.parent.toAbsolutePath().toString())
+
+    addParameter("build")
+    addParameter("--template")
+    addParameter(templatePath.toString())
+    addParameter("--build-dir")
+    addParameter(buildDir.toString())
+    if (useContainer) {
+        withParameters("--use-container")
+    }
+}
+
+fun GeneralCommandLine.samPackageCommand(
+    environmentVariables: Map<String, String>,
+    templatePath: Path,
+    packagedTemplatePath: Path,
+    s3Bucket: String
+) = this.apply {
+    withEnvironment(environmentVariables)
+    withWorkDirectory(templatePath.parent.toAbsolutePath().toString())
+
+    addParameter("package")
+    addParameter("--template-file")
+    addParameter(templatePath.toString())
+    addParameter("--output-template-file")
+    addParameter(packagedTemplatePath.toString())
+    addParameter("--s3-bucket")
+    addParameter(s3Bucket)
+}
+
+fun GeneralCommandLine.samDeployCommand(
+    environmentVariables: Map<String, String>,
+    stackName: String,
+    templatePath: Path,
+    parameters: Map<String, String>,
+    capabilities: List<CreateCapabilities>,
+    s3Bucket: String
+) = this.apply {
+    withEnvironment(environmentVariables)
+    withWorkDirectory(templatePath.parent.toAbsolutePath().toString())
+
+    addParameter("deploy")
+    addParameter("--template-file")
+    addParameter(templatePath.toString())
+    addParameter("--stack-name")
+    addParameter(stackName)
+    addParameter("--s3-bucket")
+    addParameter(s3Bucket)
+
+    if (capabilities.isNotEmpty()) {
+        addParameter("--capabilities")
+        addParameters(capabilities.map { it.capability })
+    }
+
+    addParameter("--no-execute-changeset")
+
+    if (parameters.isNotEmpty()) {
+        addParameter("--parameter-overrides")
+        parameters.forEach { (key, value) ->
+            addParameter("\"$key\"=\"$value\"")
```

Comment:
Some additional things I'm concerned about:
What if the value already has a `"`, what if it already has an escaped quote `\"`, what if the value is already quoted, what if the value is already single quoted, what if the value has a single quote in it

Reply:
Is there validation before this point? I would suggest throwing an error if the key or value contains any of `\`, `"`, or `=` (and probably more).

Reply:
I am not sure about '=', but \" is supported by SAM

https://github.com/wchengru/aws-sam-cli/blob/develop/tests/unit/cli/test_types.py#L65

Reply:
Okay...I have come to the conclusion that getting this right is impossible due to bugs in SAM:

Expected: `Hello\ World`
| Tried                              | Result        |
|------------------------------------|---------------|
| `"Hello\ World"`    | `Hello World`   |
| `"Hello\\ World"`  | `Hello World`   |
| `"Hello\\\ World"` | `Hello\\ World` |

And `Hello\World` will always become `Hello\\World`

I have also have had to switch to single quotes if already quoted due a bug where they will strip off all surrounding quotes, not just the outside ones. (This requires SAM CLI 1.1.0...)

SAM is happy with `=` it seems:
`ParameterValue=\"Hello=World\" -> `'Hello=World'`

Reply:
Can you make the box validate that? Quoting is way more complicated than it has any right to be

Reply:
I am not sure we should, as far as a I can tell it is perfectly valid in CFN to do it.

I got the quotes issues all fixed but backslashes are broken now. I plan on cutting an issue to them and seeing what they have to say.